onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc

```diff
@@ -273,48 +315,72 @@ common::Status NnapiExecutionProvider::Compile(const std::vector<onnxruntime::No
 
-          if (model_output_type.GetOperandBlobByteSize() == 0) {
-            return Status(common::ONNXRUNTIME, common::FAIL,
-                          "We do not support dynamic output shape or empty output for now");
+          void* output_buffer = nullptr;
+          if (!is_dynamic_shape_output) {
+            ORT_RETURN_IF_ERROR(GetOutputBuffer(ort, context,
+                                                *model,
+                                                output_name, output_shape, model_output_type.type,
+                                                &output_buffer));
+          } else {
+            // This output is dynamic (size unknown), will need allocate a buffer for the result
+            // and copy the content to ORT output tensors afte the execuctio (will know output shape after the execution)
+            size_t output_buffer_byte_size = model->GetDynamicOutputBufferSize() * model_output_type.GetElementByteSize();
+            std::unique_ptr<uint8_t[]> buffer_holder(new uint8_t[output_buffer_byte_size]);
+            output_buffer = buffer_holder.get();
+            dynamic_shape_output_types.push_back(model_output_type);
+            dynamic_shape_output_indices.push_back(static_cast<int32_t>(i));
+            dynamic_shape_output_buffers.push_back(std::move(buffer_holder));
           }
 
           outputs.push_back({output_buffer, std::move(model_output_type)});
         }
 
-        model->SetOutputBuffers(outputs);
-        model->Predict();
+        ORT_RETURN_IF_ERROR(model->SetOutputBuffers(outputs));
+        std::vector<std::vector<uint32_t>> dynamic_output_shapes;
+        ORT_RETURN_IF_ERROR(
+            model->Predict(dynamic_shape_output_indices, dynamic_output_shapes));
+
+        // We have dynamic output buffers, need to copy the content from temp buffers to ORT output tensors
+        for (size_t i = 0; i < dynamic_shape_output_indices.size(); i++) {
+          const int32_t model_output_idx = dynamic_shape_output_indices[i];
+          const auto output_name = model_outputs[model_output_idx];
+
+          const auto& output_shape = dynamic_output_shapes[i];
+          auto model_output_type = dynamic_shape_output_types[i];
+          model_output_type.SetDimensions(output_shape);
+          ORT_RETURN_IF_NOT(model_output_type.GetOperandBlobByteSize() != 0, "We do not support 0 size output for now");
+
+          void* model_output_buffer = dynamic_shape_output_buffers[i].get();
+          void* onnx_output_buffer = nullptr;
+          ORT_RETURN_IF_ERROR(GetOutputBuffer(ort, context,
+                                              *model,
+                                              output_name, output_shape, model_output_type.type,
+                                              &onnx_output_buffer));
+
+          size_t output_buffer_byte_size = model_output_type.GetOperandBlobByteSize();
+          memcpy(onnx_output_buffer, model_output_buffer, output_buffer_byte_size);
+        }
```

Comment:
Any chance output_buffer_byte_size could be larger than model_output_buffer?

Reply:
It should not, if this is the case ( the output overflows the buffer), the execution will fail and return onnx_output_buffer

---
In reply to: [462112636](https://github.com/microsoft/onnxruntime/pull/4650#discussion_r462112636) [](ancestors = 462112636)